cpp/src/parquet/encoding.cc

```diff
@@ -1559,11 +1631,19 @@ class DictDecoderImpl : public DecoderImpl, virtual public DictDecoder<Type> {
         valid_bits, valid_bits_offset, num_values, null_count,
         [&]() { valid_bytes[i++] = 1; }, [&]() { ++i; });
 
-    auto binary_builder = checked_cast<::arrow::BinaryDictionary32Builder*>(builder);
-    PARQUET_THROW_NOT_OK(
-        binary_builder->AppendIndices(indices_buffer, num_values, valid_bytes.data()));
-    num_values_ -= num_values - null_count;
-    return num_values - null_count;
+    // It looks like this method is only called by ByteArray types. Previously,
+    // there was an unconditional cast to ::arrow::Dictionary32Builder<::arrow::BinaryType>.
+    // This won't work for LargeByteArrayType and the Type template argument can't be used
+    // unconditionally because it is not defined for several other types.
+    if constexpr (std::is_same_v<ByteArrayType, Type> || std::is_same_v<LargeByteArrayType, Type>) {
+      auto binary_builder = checked_cast<typename EncodingTraits<Type>::DictAccumulator*>(builder);
+      PARQUET_THROW_NOT_OK(
+          binary_builder->AppendIndices(indices_buffer, num_values, valid_bytes.data()));
+      num_values_ -= num_values - null_count;
+      return num_values - null_count;
+    }
+
+    ParquetException::NYI("DecodeIndicesSpaced not implemented for this type");
```

Comment:
@mapleFU @wgtmac Please share your thoughts on this

Reply:
Yes, currently only ByteArray type supports directly reading indices of dictionary. See below for evidence:

https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_reader.cc#L2230-L2242

This change should be fine for now. But I am not sure if we will support reading dictionary indices from other types in the future.

Reply:
I guess report error when detect mismatched type is ok.

```c++
/// \brief Read records directly to dictionary-encoded Arrow form (int32
/// indices). Only valid for BYTE_ARRAY columns
class DictionaryRecordReader : virtual public RecordReader {
 public:
  virtual std::shared_ptr<::arrow::ChunkedArray> GetResult() = 0;
};
```

I'm not sure I understand it. The dictionary-encoded Arrow form should still has int32 dictionary index?

Reply:
I am not sure I understand your question

Reply:
https://arrow.apache.org/docs/format/Columnar.html#dictionary-encoded-layout @arthurpassos 
Even if it's large binary, the "dictionary index" is 32bits width. It's apart from 64bits offset.

Reply:
Yes, this PR only addresses overflow of concatenated string length, not the number of dictionary entries. So here we still keep using int32_t for dictionary indices only.

Reply:
Ok, so there is no problem, right? We are using the 32 bit form as expected.

Reply:
Oh yeah, `Dictionary32Builder<LargeBinaryType>` uses int32